designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -151,13 +167,25 @@ class DeploymentServiceImpl(dispatcher: DeploymentManagerDispatcher,
   }
 
   override def getProcessState(processIdWithName: ProcessIdWithName)
-                              (implicit user: LoggedUser, ec: ExecutionContext): Future[ProcessState] =
+                              (implicit user: LoggedUser, ec: ExecutionContext): Future[ProcessState] = {
     for {
-      actions <- processRepository.fetchProcessActions(processIdWithName.id)
       manager <- dispatcher.deploymentManager(processIdWithName.id)
-      state <- findJobState(manager, processIdWithName)
-      _ <- handleFinishedProcess(processIdWithName, state)
-    } yield ObsoleteStateDetector.handleObsoleteStatus(state, actions.headOption)
+      inProgressActionTypes <- deploymentActionsInProgressRepository.getDeploymentActionInProgressTypes(processIdWithName)
+      result <- {
+        if (inProgressActionTypes.contains(Deployment)) {
+          Future.successful(manager.processStateDefinitionManager.processState(SimpleStateStatus.DuringDeploy))
+        } else if (inProgressActionTypes.contains(Cancel)) {
+          Future.successful(manager.processStateDefinitionManager.processState(SimpleStateStatus.DuringCancel))
+        } else {
+          for {
+            actions <- processRepository.fetchProcessActions(processIdWithName.id)
+            state <- findJobState(manager, processIdWithName)
+            _ <- handleFinishedProcess(processIdWithName, state)
+          } yield ObsoleteStateDetector.handleObsoleteStatus(state, actions.headOption)
+        }
+      }
+    } yield result
+  }
```

Comment:
will it work correctly if Designer has downtime between deploy and ivocation of this method?

Reply:
in the mean time scenario could fail, or finish..
and hadnling finished scenario's is in `else` branch

Reply:
Regarding correctness - it will work  exactly as it work now, this is only a refactor, nothing changed. Are you asking about further plans? Before this change many things was done outside of Actor context because they were done in Future.